File: ports/elixir/lib/unicode_security/form/nfc_idempotence_witness.ex
defmodule UnicodeSecurity.Form.NfcIdempotenceWitness do
  alias UnicodeSecurity.Ucd

  def detect(input) do
    cond do
      (pos = first_divergence(input, Ucd.to_nfc(input))) != nil ->
        %{sub: "NonNfcForm", positions: [pos]}

      (pos = first_divergence(input, Ucd.to_nfkc(input))) != nil ->
        %{sub: "NonNfkcCompatForm", positions: [pos]}

      true ->
        %{sub: nil, positions: []}
    end
  end

  defp first_divergence(a, b) do
    n = min(length(a), length(b))
    pos = Enum.find(0..max(n - 1, 0), fn i -> n > 0 and Enum.at(a, i) != Enum.at(b, i) end)

    cond do
      pos != nil -> pos
      length(a) != length(b) -> n
      true -> nil
    end
  end
end

File: ports/php/src/Security/Form/NfcIdempotenceWitness.php
<?php

declare(strict_types=1);

namespace UnicodePhp\Security\Form;

use UnicodePhp\Security\Identity\Ucd;

final class NfcIdempotenceWitness
{
    /** @param list<int> $a @param list<int> $b */
    private static function firstDivergence(array $a, array $b): ?int
    {
        $n = min(count($a), count($b));
        for ($i = 0; $i < $n; $i++) {
            if ($a[$i] !== $b[$i]) {
                return $i;
            }
        }
        return count($a) === count($b) ? null : $n;
    }

    /** @param list<int> $input */
    public static function detect(array $input): FormVerdict
    {
        $pos = self::firstDivergence($input, Ucd::toNfc($input));
        if ($pos !== null) {
            return new FormVerdict('NonNfcForm', [$pos]);
        }
        $pos = self::firstDivergence($input, Ucd::toNfkc($input));
        if ($pos !== null) {
            return new FormVerdict('NonNfkcCompatForm', [$pos]);
        }
        return new FormVerdict(null, []);
    }
}
